doc: caf: Add description of what KEY_ID is (#6451)
nrfconnect/sdk-nrf: doc/nrf/libraries/caf/buttons.rst

Whenever a button is pressed, the module switches to ``STATE_SCANNING``.
When the switch occurs, the module submits a work with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_DEBOUNCE_INTERVAL`.
The work scans the keyboard matrix, or directly connected buttons (depends on configuration).

> If any button state change occurs, the module sends event with :c:member:`button_event.key_id` of that button.

Author: done.

* If the button is kept pressed while the scanning is performed, the work will be resubmitted with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_SCAN_INTERVAL`.
* If no button is pressed, the module switches back to ``STATE_ACTIVE``.

Key ID

> ======

Reviewer: Please add a blank line after this line.

Author: done.

The :c:member:`button_event.key_id` is an unique 2-byte value that depends on the module configuration described in the configuration file.

In this :c:member:`button_event.key_id` value:

> * Most significant byte is the index of the ``col`` array that corresponds to the gpio port and the pin to which the button's column is connected.

Reviewer: Please add "The" at the start of this and the following line. Thanks.

Author: done.

> Key ID
> ======
> 
> The :c:member:`button_event.key_id` is an unique 2 Byte value that depends on module configuration enclosed in the configuration file.
> 
> * Most significant byte - index of ``col`` array that corresponds to gpio port and pin to which button's column is connected.
> * Least significant byte - index of ``row`` array that corresponds to gpio port and pin to which button's row is connected.
> 
> To give an example let's assume your configuration file looks as in `Configuration`_ section of this documentation.
> An :c:struct:`button_event` with :c:member:`button_event.key_id` value equal to `0x0103` means that:
> 
> * The button column pin is connected to gpio that is at index 1 in ``col`` array from the configuration file.
>   That is column port and pin are 0 and 24 respectively.
> * The button row pin is connected to gpio that is at index 3 in ``col`` array from the configuration file.
>   That is row port and pin are 1 and 14 respectively.
> 

Reviewer: ```suggestion
Key ID
======

The :c:member:`button_event.key_id` is an unique 2-byte value that depends on the module configuration described in the configuration file.

In this :c:member:`button_event.key_id` value:

* Most significant byte is the index of the ``col`` array that corresponds to the gpio port and the pin to which the button's column is connected.
* Least significant byte is the index of the ``row`` array that corresponds to the gpio port and the pin to which the button's row is connected.

For example, if the configuration file looks as described in the `Configuration`_ section of this documentation, an event :c:struct:`button_event` with the :c:member:`button_event.key_id` value equal to ``0x0103`` indicates the following information:

* The button column pin is connected to the gpio that is at index 1 in the ``col`` array from the configuration file.
  That is, the column port and pin are ``0`` and ``24``, respectively.
* The button row pin is connected to the gpio that is at index 3 in the ``col`` array from the configuration file.
  That is, the row port and pin are ``1`` and ``14``, respectively.
```

Author: done

Whenever a button is pressed, the module switches to ``STATE_SCANNING``.
When the switch occurs, the module submits a work with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_DEBOUNCE_INTERVAL`.
The work scans the keyboard matrix, or directly connected buttons (depends on configuration).
If any button state change occurs, the module sends an event with the :c:member:`button_event.key_id` of that button.

* If the button is kept pressed while the scanning is performed, the work will be resubmitted with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_SCAN_INTERVAL`.
* If no button is pressed, the module switches back to ``STATE_ACTIVE``.

Key ID
======

The :c:member:`button_event.key_id` is an unique 2-byte value that depends on the module configuration described in the configuration file.

In this :c:member:`button_event.key_id` value:

> * The most significant byte is the index of the ``col`` array that corresponds to the gpio port and the pin to which the button's column is connected.

Reviewer 2: That's unfortunately not true.
Looking from the least significant bit:
7 bits represent row, 7 bits represent columnt, 1 bit represents if this is function key (feature currently specific to nRF desktop). See `nrf/include/caf/key_id.h` for details.

Author: Mea culpa. I didn't check it well enough. I will rewrite this, but i will avoid nrf_desktop specific information in this documentation. I will add a note to `nrf_desktop/doc/buttons.rst` about that. Is that okay for you?

Whenever a button is pressed, the module switches to ``STATE_SCANNING``.
When the switch occurs, the module submits a work with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_DEBOUNCE_INTERVAL`.
The work scans the keyboard matrix, or directly connected buttons (depends on configuration).
If any button state change occurs, the module sends an event with the :c:member:`button_event.key_id` of that button.

* If the button is kept pressed while the scanning is performed, the work will be resubmitted with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_SCAN_INTERVAL`.
* If no button is pressed, the module switches back to ``STATE_ACTIVE``.

Key ID
======

The :c:member:`button_event.key_id` is an unique 2-byte value that depends on the module configuration described in the configuration file.

In this :c:member:`button_event.key_id` value:

> * The first seven least significant bits encodes the index of the ``row`` array that corresponds to the gpio port and the pin to which the button's row is connected.

Reviewer 2: Consider using an array (in similar way as we describe data format here: https://raw.githubusercontent.com/nrfconnect/sdk-nrf/main/applications/nrf_desktop/doc/config_channel.rst).
That could be more readable. Discussed offline.

Whenever a button is pressed, the module switches to ``STATE_SCANNING``.
When the switch occurs, the module submits a work with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_DEBOUNCE_INTERVAL`.
The work scans the keyboard matrix, or directly connected buttons (depends on configuration).
If any button state change occurs, the module sends an event with the :c:member:`button_event.key_id` of that button.

* If the button is kept pressed while the scanning is performed, the work will be resubmitted with a delay set to :kconfig:`CONFIG_CAF_BUTTONS_SCAN_INTERVAL`.
* If no button is pressed, the module switches back to ``STATE_ACTIVE``.

Key ID
======

The :c:member:`button_event.key_id` is an unique 2-byte value that depends on the module configuration described in the configuration file.

+------------------------------+
| key_id                       |
+---------+---------+----+-----+
| 0 .. 6  | 7 .. 13 | 14 | 15  |
+=========+=========+====+=====+
| Row idx | Col idx | Reserved |

> +---------+---------+----------+

Reviewer: What does this table show?
We need to add an introductory statement before it.
```suggestion
The following table breaks down the ``key_id`` bits.

+------------------------------+
| key_id                       |
+---------+---------+----+-----+
| 0 .. 6  | 7 .. 13 | 14 | 15  |
+=========+=========+====+=====+
| Row idx | Col idx | Reserved |
+---------+---------+----------+

In this :c:member:`button_event.key_id` value:

> * The first seven least significant bits (``Row idx``) encodes the index of the ``row`` array that corresponds to the gpio port and the pin to which the button's row is connected.

Reviewer: ```suggestion
* The first seven least significant bits (``Row idx``) encode the index of the ``row`` array that corresponds to the GPIO port and the pin to which the button's row is connected.

> * The next seven bits (``Col idx``) encodes the index of the ``col`` array that corresponds to the gpio port and the pin to which the button's column is connected.

Reviewer 2: You could mention that in case of buttons that are directly connected (no keyboard matrix), the buttons are denoted only by row and col is always set to zero.

> * The next seven bits (``Col idx``) encodes the index of the ``col`` array that corresponds to the gpio port and the pin to which the button's column is connected.

Reviewer: ```suggestion
* The next seven bits (``Col idx``) encode the index of the ``col`` array that corresponds to the GPIO port and the pin to which the button's column is connected.

> * The last two bits are reserved.

Reviewer: ```suggestion
* The last two bits are reserved for application-specific usage.

For example, if the configuration file looks as described in the `Configuration`_ section of this documentation, an event :c:struct:`button_event` with the :c:member:`button_event.key_id` value equal to ``0x0103`` indicates the following information:

> * The button column pin is connected to the gpio that is at index 1 in the ``col`` array from the configuration file.

Reviewer 2: Add additional column too, because we are getting out of array bonds ;)
